File: java/src/main/java/neetcode/twopointers/p0015_threesum/Solution.java
package neetcode.twopointers.p0015_threesum;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Solution {
    /**
     * Returns all unique triplets [a, b, c] in nums such that a + b + c = 0.
     *
     * Approach: sort the array, then for each index i, use converging
     * two pointers (left, right) on the right portion to find pairs that
     * sum to -nums[i]. Skip duplicates at both the outer and inner levels.
     *
     * Time:  O(n^2) — outer loop times inner two-pointers walk
     * Space: O(1) extra (excluding output and the cost of sorting)
     */
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        Arrays.sort(nums);

        int n = nums.length;
        for (int i = 0; i < n - 2; i++) {
            // With a sorted array, once nums[i] > 0 no further triplet can sum to 0
            if (nums[i] > 0) break;
            // Skip duplicate values at the outer level
            if (i > 0 && nums[i] == nums[i - 1]) continue;

            int left = i + 1;
            int right = n - 1;

            while (left < right) {
                int sum = nums[i] + nums[left] + nums[right];

                if (sum == 0) {
                    result.add(Arrays.asList(nums[i], nums[left], nums[right]));
                    // Skip duplicates on the left side
                    while (left < right && nums[left] == nums[left + 1]) left++;
                    // Skip duplicates on the right side
                    while (left < right && nums[right] == nums[right - 1]) right--;
                    left++;
                    right--;
                } else if (sum < 0) {
                    left++;     // sum too small → need a bigger left
                } else {
                    right--;    // sum too big → need a smaller right
                }
            }
        }

        return result;
    }
}

File: python/src/neetcode/twopointers/p0015_threesum/solution.py
from typing import List


class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        """
        Returns all unique triplets that sum to zero.

        Approach: sort + outer loop + inner two pointers.
        Sort enables two-pointer search and easy duplicate skipping.
        Skip duplicate values at the outer level (i) and inner level
        (left/right) to avoid repeated triplets in the output.

        Time:  O(n^2) — outer loop O(n) * inner two pointers O(n)
        Space: O(1) extra (excluding output)
        """
        nums.sort()
        result = []

        for i in range(len(nums) - 2):
            if nums[i] > 0:
                break                          # all remaining sums will be positive
            if i > 0 and nums[i] == nums[i - 1]:
                continue                       # skip duplicate outer values

            left, right = i + 1, len(nums) - 1
            while left < right:
                s = nums[i] + nums[left] + nums[right]
                if s == 0:
                    result.append([nums[i], nums[left], nums[right]])
                    while left < right and nums[left] == nums[left + 1]:
                        left += 1              # skip duplicate left values
                    while left < right and nums[right] == nums[right - 1]:
                        right -= 1             # skip duplicate right values
                    left += 1
                    right -= 1
                elif s < 0:
                    left += 1
                else:
                    right -= 1

        return result
